sdk/src/azure/platform/az_mosquitto.c

```diff
@@ -0,0 +1,278 @@
+static void _az_mosquitto_on_publish(struct mosquitto* mosq, void* obj, int mid)
+{
+  (void)mosq;
+  az_mqtt* me = (az_mqtt*)obj;
+
+  az_result ret = az_mqtt_inbound_puback(me, &(az_mqtt_puback_data){ mid });
+
+  if (az_result_failed(ret))
+  {
+    _az_mosquitto_critical_error();
+  }
+}
+
+static void _az_mosquitto_on_subscribe(
+    struct mosquitto* mosq,
+    void* obj,
+    int mid,
+    int qos_count,
+    const int* granted_qos)
+{
+  (void)mosq;
+  (void)qos_count;
+  (void)granted_qos;
+
+  az_mqtt* me = (az_mqtt*)obj;
+
+  az_result ret = az_mqtt_inbound_suback(me, &(az_mqtt_suback_data){ mid });
+
+  if (az_result_failed(ret))
+  {
+    _az_mosquitto_critical_error();
+  }
+}
+
+static void _az_mosquitto_on_unsubscribe(struct mosquitto* mosq, void* obj, int mid)
+{
+  (void)mosq;
+  (void)obj;
+  (void)mid;
+
+  // Unsubscribe is not handled or implemented. We should never get here.
+  _az_mosquitto_critical_error();
+}
+
+static void _az_mosquitto_on_message(
+    struct mosquitto* mosq,
+    void* obj,
+    const struct mosquitto_message* message)
+{
+  (void)mosq;
+  az_mqtt* me = (az_mqtt*)obj;
+
+  az_result ret = az_mqtt_inbound_recv(
+      me,
+      &(az_mqtt_recv_data){ .qos = (int8_t)message->qos,
+                            .id = (int32_t)message->mid,
+                            .payload = az_span_create(message->payload, message->payloadlen),
+                            .topic = az_span_create_from_str(message->topic) });
+
+  if (az_result_failed(ret))
+  {
+    _az_mosquitto_critical_error();
+  }
+}
+
+static void _az_mosquitto_on_log(struct mosquitto* mosq, void* obj, int level, const char* str)
+{
+  (void)mosq;
+  (void)obj;
+  (void)level;
+  if (_az_LOG_SHOULD_WRITE(AZ_LOG_MQTT_STACK))
+  {
+    _az_LOG_WRITE(AZ_LOG_MQTT_STACK, az_span_create_from_str((char*)(uintptr_t)str));
+  }
+}
+
+AZ_NODISCARD az_mqtt_options az_mqtt_options_default()
+{
+  return (az_mqtt_options){ .certificate_authority_trusted_roots = NULL,
+                            .openssl_engine = NULL,
+                            .mosquitto_handle = NULL };
+}
+
+AZ_NODISCARD az_result az_mqtt_init(az_mqtt* mqtt, az_mqtt_options const* options)
+{
+  _az_PRECONDITION_NOT_NULL(mqtt);
+  mqtt->_internal.options = options == NULL ? az_mqtt_options_default() : *options;
+  mqtt->mosquitto_handle = mqtt->_internal.options.mosquitto_handle;
+  mqtt->_internal.platform_mqtt.pipeline = NULL;
+
+  return AZ_OK;
+}
+
+AZ_NODISCARD az_result
+az_mqtt_outbound_connect(az_mqtt* mqtt, az_context* context, az_mqtt_connect_data* connect_data)
+{
+  (void)context;
+  az_result ret = AZ_OK;
+  az_mqtt* me = (az_mqtt*)mqtt;
+
+  // IMPORTANT: application must call mosquitto_lib_init() before any Mosquitto clients are created.
+
+  if (me->mosquitto_handle == NULL)
+  {
+    me->mosquitto_handle = mosquitto_new(
+        (char*)az_span_ptr(connect_data->client_id),
+        false, // clean-session
+        me); // callback context - i.e. user data.
```

Comment:
Related to the question above - Isn't this sort of recursively containing `me`? Since then the mosquitto mosq holds me in the obj, which holds the mosquitto mosq?

Reply:
Not sure I understand what you are referring to as "obj". This is attaching our az_mqtt object to the Mosquitto handle so that we can retrieve it later in the callback. 

Maybe this might clarify things a bit more: we support any number of `az_mqtt` objects within the application context. This means that the callbacks can occur at the same time, on different threads, for different `az_mqtt`/mosquitto objects. We need to ensure we are operating on the correct pipeline, and that is done by passing the equivalent of the `this` pointer (`me` here) into the context.

Reply:
obj is the third parameter of mosquitto_new: "A user pointer that will be passed as an argument to any callbacks that are specified."
We get it back from mosquitto in functions like `static void _az_mosquitto_on_disconnect(struct mosquitto* mosq, void* obj, int rc)`

It makes sense that the az_mqtt would contain the mosquitto_handle because of threads/different implementations, but I don't think we should be passing in the mosquitto_handle as part of the obj for the mosquitto implemetation, since every callback also gives us the mosquitto handle as the `mosq` parameter.

For mosquitto, the `mosquitto_handle` (returned on `mosquitto_new`), contains the obj (`me` for our wording), accessible with [mosquitto_userdata](https://mosquitto.org/api/files/mosquitto-h.html#mosquitto_userdata). But because we put `me` in the obj, it's a little weird where
`mosquitto_userdata(me->mosquitto_handle)` returns me, which you could then use to get the userdata on that mosquitto_handle property of that value and get `me` again.

Let me know if this doesn't clarify things!